cirq/google/calibration/engine_simulator.py

```diff
@@ -0,0 +1,405 @@
+from typing import (
+    Any,
+    Callable,
+    Dict,
+    Iterable,
+    Iterator,
+    List,
+    Optional,
+    Sequence,
+    Tuple,
+    Union,
+    cast,
+)
+
+import numpy as np
+
+from cirq.circuits import Circuit, PointOptimizer, PointOptimizationSummary
+from cirq.ops import (
+    FSimGate,
+    Gate,
+    MeasurementGate,
+    Operation,
+    PhasedFSimGate,
+    Qid,
+    QubitOrderOrList,
+    SingleQubitGate,
+    WaitGate,
+)
+from cirq.sim import (
+    Simulator,
+    SimulatesSamples,
+    SimulatesIntermediateStateVector,
+    SparseSimulatorStep,
+    StepResult,
+)
+from cirq.study import ParamResolver
+from cirq.value import RANDOM_STATE_OR_SEED_LIKE, parse_random_state
+
+from cirq.google.calibration.phased_fsim import (
+    FloquetPhasedFSimCalibrationRequest,
+    IncompatibleMomentError,
+    PhasedFSimCalibrationRequest,
+    PhasedFSimCalibrationResult,
+    PhasedFSimCharacterization,
+    SQRT_ISWAP_PARAMETERS,
+    try_convert_sqrt_iswap_to_fsim,
+)
+
+
+ParametersDriftGenerator = Callable[[Qid, Qid, FSimGate], PhasedFSimCharacterization]
+PhasedFsimDictParameters = Dict[
+    Tuple[Qid, Qid], Union[Dict[str, float], PhasedFSimCharacterization]
+]
+
+
+class PhasedFSimEngineSimulator(SimulatesSamples, SimulatesIntermediateStateVector):
+    """Wrapper on top of cirq.Simulator that allows to simulate calibration requests.
+
+    This simulator introduces get_calibrations which allows to simulate
+    cirq.google.run_characterizations requests. The returned calibration results represent the
+    internal state of a simulator. Circuits which are run on this simulator are modified to account
+    for the changes in the unitary parameters as described by the calibration results.
+    """
+
+    def __init__(
+        self,
+        simulator: Simulator,
```

Comment:
If I set the default argument then I either have to make some dummy drift_generator as well, or I move simulator as a second argument following the drift_generator. I wouldn't want to do that because I want to keep the simulator non-keyword argument. My intention was that this initializer is actually private, and creation happens inside.

Reply:
works for me. I see most people will use the factory functions